### ISSUE TITLE
JPA monitoring and hibernate-jpa2.1-api

### ISSUE THREAD

**Issue by lelmarir** [javamelody/javamelody#869]
Hi,

I've found two problems in the actual implementation:
- if Javamelody is present in the classpath, it will be used as provider (if appen to be the first provider among discovered ones), no mater wich provider is set in persistence.xml 
- Hibernate Delegate will not inizialize if provider is set to javamelody one

setting `<provider>net.bull.javamelody.JpaPersistence</provider>` and `<property name="net.bull.javamelody.jpa.provider" value="org.hibernate.jpa.HibernatePersistenceProvider" />`
is not guarantee to always work:

in **javax.persistence.Persistence.java** (hibernate-jpa-2.1-api 1.0.2.Final)

	public static EntityManagerFactory createEntityManagerFactory(String persistenceUnitName, Map properties) {
		EntityManagerFactory emf = null;
		List<PersistenceProvider> providers = getProviders();
		for ( PersistenceProvider provider : providers ) {
			emf = provider.createEntityManagerFactory( persistenceUnitName, properties );
			if ( emf != null ) {
				break;
			}
		}
		if ( emf == null ) {
			throw new PersistenceException( "No Persistence provider for EntityManager named " + persistenceUnitName );
		}
		return emf;
	}

Let suppose that the real Hibernate provider is cecked first (conversely is the same), it will see tha it is not the provider for the unit and will continue:

**org.hibernate.jpaHibernatePersistenceProvider**

	// See if we (Hibernate) are the persistence provider
	if ( ! ProviderChecker.isProvider( persistenceUnit, properties ) ) {
		log.debug( "Excluding from consideration due to provider mis-match" );
		continue;
	}

instead the Javamelody provider will ignoire the setted provider and proceed creating the delegate

**net.bull.javamelody.JpaPersistence**

	public EntityManagerFactory createEntityManagerFactory(final String unit, final Map map) {
		//...
		final PersistenceProvider persistenceProvider = findDelegate(map);
		//...
			final EntityManagerFactory entityManagerFactory = persistenceProvider
					.createEntityManagerFactory(unit, map);
		//...
	}

the provider is correctly identified (hibernate) but `persistenceProvider.createEntityManagerFactory` will fail to create the provider because hibernate will ceck if he is the provider, but he is not (the provider is net.bull.javamelody.JpaPersistence)

A way to fix this would be:
- let javamelody provider ceck if he is the provider before proceding (like hibernate do)
- let javamelody inject `javax.persistence.provider=org.hibernate.jpa.HibernatePersistenceProvider` (as default, or any value specified in `net.bull.javamelody.jpa.provider`) into the map before call persistenceProvider.createEntityManagerFactory, this way hiberneate will tink that he is te provider.

The only way i've found to enable javalemody is to not set any provider in persistence.xml,
this way Javamelody will create the delagate and hibernate will speculate he is the provider... 
but this will work only if Javamelody is checked before hibernate among avalible providers (in a HashSet)....

**Comment by evernat** [comment]
> if Javamelody is present in the classpath, it will be used as provider (if appen to be the first provider among discovered ones), no mater wich provider is set in persistence.xml

This is not a problem. It is supposed to be a feature. The goal is to plug the JPA monitoring, if possible, even if `<provider>net.bull.javamelody.JpaPersistence</provider>` is not added in persistence.xml.

> let javamelody provider ceck if he is the provider before proceding (like hibernate do)

It probably won't be done, because the behavior above is a feature.

> Hibernate Delegate will not inizialize if provider is set to javamelody one

This is a problem indeed. The Hibernate delegate is supposed to be initialized and used. I have reproduced the exception using Hibernate 5.4:
```
javax.persistence.PersistenceException: No Persistence provider for EntityManager named null
	at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:85)
	at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:54)
```

In javamelody [JpaPersistence](https://github.com/javamelody/javamelody/blob/master/javamelody-core/src/main/java/net/bull/javamelody/JpaPersistence.java#L102), the persistence.xml loaded by Hibernate and overridden by [JpaOverridePersistenceXmlClassLoader](https://github.com/javamelody/javamelody/blob/master/javamelody-core/src/main/java/net/bull/javamelody/JpaOverridePersistenceXmlClassLoader.java#L38) should not have the provider declared initially and Hibernate should be ok.
But this does not work anymore, because Hibernate now uses the system classloader before the Thread Context Class Loader (TCCL). I think that [this PR](https://github.com/hibernate/hibernate-orm/pull/1636/files) caused the change: it was TCCL [before](https://github.com/hibernate/hibernate-orm/pull/1636/files#diff-ba8e26dbd489d03922fe26f46ceac9efL82) system CL, now it is TCCL [after](https://github.com/hibernate/hibernate-orm/pull/1636/files#diff-ba8e26dbd489d03922fe26f46ceac9efR113) system CL.

Injecting the (Hibernate) persistence provider into the map before calling createEntityManagerFactory seems to work for Hibernate. To be fixed...

**Comment by evernat** [comment]
The `PersistenceException: No Persistence provider for EntityManager` issue, with Hibernate 5.4 and JPA is fixed by bc1bab6 and ready for the next release. (Given that JpaOverridePersistenceXmlClassLoader was not enough in this case, the persistence provider is injected into the map before calling createEntityManagerFactory.)

Moreover the same commit always enables the JPA monitoring, even if `<provider>net.bull.javamelody.JpaPersistence</provider>` is not added in persistence.xml. This is done using the JPA 2 resolver api. This is improved compared to previous behavior, which needed the chance to have javamelody as the first persistence provider or to have the javamelody provider in persistence.xml.

A snapshot build is available at: 
http://javamelody.org/downloads/javamelody-core-1.81.0-SNAPSHOT.jar

**Comment by evernat** [comment]
Note that spring-data-jpa does not seem to use `Persistence#createEntityManagerFactory` of JPA. So JPA is not automatically monitored in spring-data-jpa.
